std/internal/mmm.d

```diff
@@ -0,0 +1,337 @@
+    scope(failure) () @trusted { free(p); }();
+
+    emplaceRef(*p, args);
+    return p;
+}
+
+/**
+ * Copy the elements of range into a new array allocated with $(D malloc).
+ *
+ * Params:
+ *    T = element type of the new array; defaults to $(D ElementType!Range).
+ *    range = _range of elements to copy into the new array
+ * Throws:
+ *    $(D OutOfMemoryError) on OOM
+ */
+T[] checkedMallocArray(T, Range)(Range range) // safety inferred
+    if (isInputRange!Range && hasLength!Range && is(ElementType!Range : T))
+{
+    import core.stdc.stdlib : free;
+    import std.exception : collectException;
+    import std.conv : emplaceRef;
+    import std.traits : hasElaborateDestructor;
+
+    immutable length = range.length;
+    auto arr = () @trusted { return checkedMallocArray!T(length); }();
+
+    size_t i = 0; // Number of constructed elements
+
+    scope(failure)
+    {
+        static if(hasElaborateDestructor!T)
+        {
+            for (; i != 0; --i)
+            {
+                // collectException(destroy(arr[i])); // Issue 12647
+                try
+                    destroy(arr[i]);
+                catch(Exception) {}
+            }
+        }
+        () @trusted { free(arr.ptr); }();
+    }
+
+    static if (isForwardRange!Range)
+        auto r = range.save;
+    else
+        alias r = range;
+
+    for(; !r.empty; ++i, r.popFront())
+        emplaceRef!T(arr[i], r.front);
+
+    assert(i == length);
+    return arr;
+}
+
+/// Ditto
+ElementType!Range[] checkedMallocArray(Range)(Range range)
+    if (isInputRange!Range && hasLength!Range)
+{
+    return checkedMallocArray!(ElementType!Range, Range)(range);
+}
+
+/**
+ * Allocate space for one or more objects on the C heap with $(D calloc).
+ *
+ * The new objects are zero-initialized but not default-constructed;
+ * construct them with $(XREF conv, emplace) before use.
+ *
+ * Params:
+ *    T = type of object(s)
+ *    n = number of objects to allocate space for
+ * Throws:
+ *   $(D OutOfMemoryError) on OOM
+ */
+T* checkedCalloc(T)() @trusted
+{
+    import core.exception : onOutOfMemoryError;
+
+    if (auto p = calloc(T.sizeof, 1))
+        return cast(T*)p;
+    else
+    {
+        onOutOfMemoryError();
+        assert(false);
+    }
+}
+
+/// Ditto
+T[] checkedCallocArray(T)(size_t n) @trusted
+{
+    import core.exception : onOutOfMemoryError;
+
+    if (auto p = calloc(T.sizeof, n))
+        return (cast(T*)p)[0 .. n];
+    else
+    {
+        onOutOfMemoryError();
+        assert(false);
+    }
+}
+
+/**
+ * Destroy the elements of array then $(D null) the _array reference.
+ * Any exception thrown from a destructor is silently ignored.
+ *
+ * Returns:
+ *    The _array as $(D void[]), useful for passing to a deallocation function
+ */
+void[] destroyArray(T)(ref T[] array) nothrow
+    if(hasElaborateDestructor!T)
+{
+    foreach_reverse(ref e; array)
+    {
+        try
+            destroy(e);
+        catch(Exception) {}
```

Comment:
I'm not sure what the idea was. How should throwing from a destructor affect other destructors in the same scope?